src/layer.ts

```diff
@@ -9,120 +10,146 @@ const generateID = () => {
+  MapboxGL.BackgroundLayout |
+  MapboxGL.FillLayout |
+  MapboxGL.FillExtrusionLayout |
+  MapboxGL.LineLayout |
+  MapboxGL.SymbolLayout |
+  MapboxGL.RasterLayout |
+  MapboxGL.CircleLayout
+);
+
+export interface Props {
+  id?: string;
+  type?: 'symbol' | 'line' | 'fill' | 'circle' | 'raster';
+  sourceId?: string;
+  before?: string;
+  sourceOptions?: Sources;
+  paint?: Paint;
+  layout?: Layout;
+  layerOptions?: MapboxGL.Layer;
+  children?: JSX.Element;
+}
+
+export interface Context {
+  map: MapboxGL.Map;
+}
+
+export default class Layer extends React.PureComponent<Props, void> {
+  public context: Context;
 
-  static propTypes = {
-    id: PropTypes.string,
-
-    type: PropTypes.oneOf([
-      'symbol',
-      'line',
-      'fill',
-      'circle',
-      'raster',
-    ]),
-
-    layout: PropTypes.object,
-    paint: PropTypes.object,
-    sourceOptions: PropTypes.object,
-    layerOptions: PropTypes.object,
-    sourceId: PropTypes.string,
-    before: PropTypes.string,
+  public static contextTypes = {
+    map: React.PropTypes.object
   };
 
-  static defaultProps = {
+  public static defaultProps = {
     type: 'symbol',
     layout: {},
-    paint: {},
+    paint: {}
   };
 
-  hover = [];
+  private hover: string[] = [];
 
-  id = this.props.id || `layer-${generateID()}`;
+  private id: string = this.props.id || `layer-${generateID()}`;
 
-  source = {
+  private source: Sources = {
     type: 'geojson',
     ...this.props.sourceOptions,
     data: {
       type: 'FeatureCollection',
-      features: [],
-    },
+      features: []
+    }
   };
 
-  geometry = (coordinates) => {
+  private geometry = (coordinates: number[]) => {
     switch (this.props.type) {
       case 'symbol':
       case 'circle': return {
         type: 'Point',
-        coordinates,
+        coordinates
       };
 
       case 'fill': return {
         type: coordinates.length > 1 ? 'MultiPolygon' : 'Polygon',
-        coordinates,
+        coordinates
       };
 
       case 'line': return {
         type: 'LineString',
-        coordinates,
+        coordinates
       };
 
       default: return null;
     }
-  };
+  }
 
-  feature = (props, id) => ({
+  private feature = (props: any, id: string) => ({
     type: 'Feature',
     geometry: this.geometry(props.coordinates),
-    properties: {
-      ...props.properties,
-      id,
-    },
+    properties: { ...props.properties },
+    id
   })
 
-  onClick = (evt) => {
-    const children = [].concat(this.props.children);
+  private onClick = (evt: any) => {
+    const children = ([] as any).concat(this.props.children);
     const { map } = this.context;
-    const { id } = this;
-    const features = map.queryRenderedFeatures(evt.point, { layers: [id] });
+    const features = map.queryRenderedFeatures(evt.point, { layers: [this.id] });
 
     features.forEach((feature) => {
-      const { properties } = feature;
-      const child = children[properties.id];
+      const { id } = feature;
```

Comment:
Why has `id` moved to `feature.id` instead of `feature.properties.id`. Seems dodgy